From d18a1ecfa76c4b3f8bbfe92e4c3db75b85717f55 Mon Sep 17 00:00:00 2001
From: Fedora Ninjas <grub2-owner@fedoraproject.org>
Date: Tue, 11 Mar 2025 12:49:40 +0100
Subject: [PATCH 1/3] linux_cmd: fix an issue with linux_args during Xen boot

In case of Xen boot linux_args isn't set and is set only in case of
Linux kernel boot what leads to NULL derefence during Xen boot
prepartions.

To avoid that move some parts of the code to
grub_arch_efi_linux_boot_image() and re-use argument `args` of
grub_arch_efi_linux_boot_image() instead of linux_args directly.

Considering that in the next patch grub_arch_efi_linux_boot_image_v2()
is introduced this patch could be droped.

Upstream-status: Not ready
Signed-off-by: Fedora Ninjas <grub2-owner@fedoraproject.org>
---
 grub-core/loader/arm64/linux.c | 41 +++++++++++++++-------------------
 1 file changed, 18 insertions(+), 23 deletions(-)

diff --git a/grub-core/loader/arm64/linux.c b/grub-core/loader/arm64/linux.c
index a3a193c..06995a1 100644
--- a/grub-core/loader/arm64/linux.c
+++ b/grub-core/loader/arm64/linux.c
@@ -70,7 +70,6 @@ grub_arch_efi_linux_check_image (struct linux_arch_kernel_header * lh)
 static grub_err_t
 finalize_params_linux (void)
 {
-  grub_efi_loaded_image_t *loaded_image = NULL;
   int node, retval, len;
   grub_err_t err = GRUB_ERR_NONE;
   void *fdt;
@@ -125,28 +124,6 @@ finalize_params_linux (void)
   grub_dprintf ("linux", "Installed/updated FDT configuration table @ %p\n",
                 fdt);
 
-  /* Convert command line to UCS-2 */
-  loaded_image = grub_efi_get_loaded_image (grub_efi_image_handle);
-  if (!loaded_image)
-    {
-      err = grub_error(grub_errno, "Failed to install fdt");
-      goto failure;
-    }
-
-  loaded_image->load_options_size = len =
-    (grub_strlen (linux_args) + 1) * sizeof (grub_efi_char16_t);
-  loaded_image->load_options =
-    grub_efi_allocate_any_pages (GRUB_EFI_BYTES_TO_PAGES (loaded_image->load_options_size));
-  if (!loaded_image->load_options)
-    {
-      err = grub_error(GRUB_ERR_BAD_OS, "failed to create kernel parameters");
-      goto failure;
-    }
-
-  loaded_image->load_options_size =
-    2 * grub_utf8_to_utf16 (loaded_image->load_options, len,
-                           (grub_uint8_t *) linux_args, len, NULL);
-
   return GRUB_ERR_NONE;
 
 failure:
@@ -174,7 +151,9 @@ grub_err_t
 grub_arch_efi_linux_boot_image (grub_addr_t addr, grub_size_t size, char *args,
 				int nx_supported)
 {
+  int len;
   grub_err_t retval;
+  grub_efi_loaded_image_t *loaded_image;
 
   retval = finalize_params_linux ();
   if (retval != GRUB_ERR_NONE)
@@ -182,6 +161,22 @@ grub_arch_efi_linux_boot_image (grub_addr_t addr, grub_size_t size, char *args,
 
   grub_dprintf ("linux", "linux command line: '%s'\n", args);
 
+  /* Convert command line to UCS-2 */
+  loaded_image = grub_efi_get_loaded_image (grub_efi_image_handle);
+  if (!loaded_image)
+    return grub_error(grub_errno, "Failed to install fdt");
+
+  loaded_image->load_options_size = len =
+    (grub_strlen (args) + 1) * sizeof (grub_efi_char16_t);
+  loaded_image->load_options =
+    grub_efi_allocate_any_pages (GRUB_EFI_BYTES_TO_PAGES (loaded_image->load_options_size));
+  if (!loaded_image->load_options)
+    return grub_error(GRUB_ERR_BAD_OS, "failed to create kernel parameters");
+
+  loaded_image->load_options_size =
+    2 * grub_utf8_to_utf16 (loaded_image->load_options, len,
+		            (grub_uint8_t *) args, len, NULL);
+
   retval = grub_efi_linux_boot (addr, size, handover_offset,
 				(void *)addr, nx_supported);
 
-- 
2.45.2

